File: BikeBuddy.Frontend/src/core/validations/event.ts
import * as yup from "yup"
import { BicycleType, EventStatus, EventType, PointDetails } from "../models/event/event-models"

const typeValues = Object.values(EventType).filter((value) => typeof value === 'number');

const bicycleTypeValues = Object.values(BicycleType).filter((value) => typeof value === 'number');

export const eventSchema = yup.object({
  name: yup
    .string()
    .min(5, "Минимум 5 символов")
    .max(100, "Максимум 100 символов")
    .required("Название обязательно"),
  description: yup
    .string()
    .min(20, "Минимум 20 символов")
    .max(2000, "Максимум 2000 символов")
    .required("Описание обязательно"),
  startAddress: yup
    .string()
    .required("Адрес старта обязателен"),
  endAddress: yup
    .string()
    .required("Адрес финиша обязателен"),
  startDate: yup
    .date()
    .min(new Date(), "Дата не может быть в прошлом")
    .required("Дата и время начала обязательны"),
  endDate: yup
    .date()
    .min(yup.ref('startDate'), "Время окончания должно быть после времени начала")
    .required("Дата и время окончания обязательны"),
  distance: yup
    .number()
    .min(1, "Минимальная дистанция 1 км")
    .required("Дистанция обязательна"),
  bicycleType: yup
    .number()
    .oneOf(bicycleTypeValues, "Некорректный тип велосипеда")
    .required("Тип велосипеда обязателен"),
  type: yup
    .number()
    .oneOf(typeValues, "Некорректный тип заезда")
    .required(),
  countMembers: yup
    .number()
    .min(1, "Участников не может быть ноль или меньше")
    .required("Число участников заезда обязательно"),
  // images: yup
  //   .array()
  //   .of(yup.string())
  //   .max(5, "Максимум 5 изображений")
})

export type EventFormData = yup.InferType<typeof eventSchema> & {
  points: PointDetails[]
  // mapImageFile: Blob | null; 
  status : EventStatus
  userId : string
};

File: BikeBuddy.Backend/src/BikeBuddy.Domain/Models/EventControl/Event.cs
using BikeBuddy.Domain.Models.AuthControl;
using BikeBuddy.Domain.Models.ChatControl;
using BikeBuddy.Domain.Models.EventControl.ValueObjects;

namespace BikeBuddy.Domain.Models.EventControl;

public class Event : ICreatedUpdateAt
{
    public Guid Id { get; set; }

    public string Name { get; private set; } = string.Empty;

    public string Description { get; private set; } = string.Empty;

    public EventType Type { get; private set; } = EventType.GROUP;

    public BicycleType BicycleType { get; private set; } = BicycleType.ANY;

    public int CountMembers { get; private set; }

    public int Distance { get; private set; }

    public string StartAddress { get; private set; } = string.Empty;

    public string EndAddress { get; private set; } = string.Empty;

    public DateTime StartDate { get; private set; }

    public DateTime EndDate { get; private set; }

    public EventDetails Details { get; private set; } = default!;

    public EventStatus Status { get; set; } = EventStatus.OPENED;

    public Guid? CreatedBy { get; set; }

    public DateTime CreatedAt { get; set; }

    public DateTime? UpdatedAt { get; set; }
    
    public AuthUser User { get; private set; } = default!;

    public GroupChat Chat { get; set; } = default!;

    public Event() { }

    private Event(
        Guid id,
        string name,
        string description,
        EventType type,
        BicycleType bicycleType,
        int countMembers,
        int distance,
        string startAddress,
        string endAddress,
        DateTime startDate,
        DateTime endDate,
        EventDetails details,
        EventStatus status,
        Guid createdBy)
    {
        Name = name;
        Description = description;
        Type = type;
        BicycleType = bicycleType;
        CountMembers = countMembers;
        Distance = distance;
        StartAddress = startAddress;
        EndAddress = endAddress;
        StartDate = startDate;
        EndDate = endDate;
        Details = details;
        Status = status;
        CreatedBy = createdBy;
    }

    public static Event Create(
        Guid id,
        string name,
        string description,
        EventType type,
        BicycleType bicycleType,
        int countMembers,
        int distance,
        string startAddress,
        string endAddress,
        DateTime startDate,
        DateTime endDate,
        EventDetails details,
        EventStatus status,
        Guid createdBy)
    {
        return new Event(
            id,
            name, 
            description, 
            type, 
            bicycleType, 
            countMembers,
            distance, 
            startAddress,
            endAddress, 
            startDate, 
            endDate,
            details, 
            status, 
            createdBy);
    }
}
